src/supernova/mod.rs

```diff
@@ -813,13 +815,74 @@ where
   (ck_primary, ck_secondary)
 }
 
-/// SuperNova helper trait, for implementors that provide sets of sub-circuits to be proved via NIVC.
-pub trait CircuitSet<G: Group> {
+/// SuperNova helper trait, for implementors that provide sets of sub-circuits to be proved via NIVC. `C1` must be a
+/// type (likely an `Enum`) for which a potentially-distinct instance can be supplied for each `index` below
+/// `self.num_circuits()`.
+pub trait NonUniformCircuit<G1, G2, C1>
+where
+  G1: Group<Base = <G2 as Group>::Scalar>,
+  G2: Group<Base = <G1 as Group>::Scalar>,
+  C1: StepCircuit<G1::Scalar>,
+{
   /// Initial program counter, defaults to zero.
-  fn initial_program_counter(&self) -> G::Scalar {
-    G::Scalar::ZERO
+  fn initial_program_counter(&self) -> G1::Scalar {
+    G1::Scalar::ZERO
+  }
+
+  /// Return the initial running claims for `NonUniformCircuit`'s sub-circuits.
+  fn initial_running_claims(
+    &self,
+  ) -> Vec<RunningClaim<G1, G2, C1, TrivialSecondaryCircuit<G2::Scalar>>> {
+    (0..self.num_circuits())
+      .map(|i| {
+        RunningClaim::new(
+          i,
+          self.primary_circuit(i),
+          self.secondary_circuit(),
+          self.num_circuits(),
+        )
+      })
+      .collect()
   }
 
-  /// How many augmented circuits are provided?
-  fn num_augmented_circuits(&self) -> usize;
+  /// Return digest and initial running claims.
+  fn compute_digest_and_initial_running_claims(
+    &self,
+  ) -> (
+    G1::Scalar,
+    Vec<RunningClaim<G1, G2, C1, TrivialSecondaryCircuit<G2::Scalar>>>,
+  ) {
+    let mut running_claims = self.initial_running_claims();
+
+    let running_claim_params = running_claims
+      .iter()
+      .map(|c| c.get_public_params())
+      .collect::<Vec<_>>();
+
+    let (ck_primary, ck_secondary) = compute_commitment_keys(&running_claim_params);
+
+    for claim in &mut running_claims {
+      claim.set_commitment_key(ck_primary.clone(), ck_secondary.clone());
```

Comment:
It seems quite essential to make sure that the `params` field of `RunningClaim` is an `Rc` to a `PublicParams` singleton, this sort of duplication is quickly going to be unmanageable.

Note the only reason I could see that we're manipulating a `Vec<RunningClaim<...>>` is to be able to index into them. We could have a `struct RunningClaim<...>` which would host the public parameter reference and some `impl Index`, `impl IndexMut` impls. Shall we open an issue?

Reply:
Yes, agreed — please create an issue.

Reply:
https://github.com/lurk-lab/arecibo/issues/29